File: frontend/js/views/cpi_rent_adjustment.js
// CPI rent adjustment generator — index-ratio escalation bounded by a
// floor/ceiling collar, via /calc/cpi-rent-adjustment.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
const pct = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 4 }) + '%');
let LAST_DOC = null;

export async function renderCpiRentAdjustment(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.cpirent.h1.title">// CPI RENT ADJUSTMENT</span></h1>
        <p class="muted small" data-i18n="view.cpirent.hint.intro">
            The index-based escalation used in commercial leases. Rent is reset by the ratio of a current
            consumer price index to a base-period index, bounded by an optional collar (a floor and/or ceiling
            on the percentage increase). Unlike fixed-percentage escalation, the increase tracks published CPI.
            Drafting aid, not legal advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.cpirent.h2.inputs">Adjustment inputs</h2>
            <form id="cpirent-form" class="inline-form">
                <label><span data-i18n="view.cpirent.label.state">State</span>
                    <input type="text" name="state" value="Delaware" required></label>
                <label><span data-i18n="view.cpirent.label.landlord">Landlord</span>
                    <input type="text" name="landlord_name" value=""></label>
                <label><span data-i18n="view.cpirent.label.tenant">Tenant</span>
                    <input type="text" name="tenant_name" value=""></label>
                <label><span data-i18n="view.cpirent.label.property">Premises</span>
                    <input type="text" name="property_label" value="Floor 7"></label>
                <label><span data-i18n="view.cpirent.label.base">Current annual rent ($)</span>
                    <input type="number" step="1000" min="0" name="base_rent_usd" value="50000" required></label>
                <label><span data-i18n="view.cpirent.label.cpi_base">Base CPI</span>
                    <input type="number" step="0.1" min="0" name="cpi_base" value="280.0" required></label>
                <label><span data-i18n="view.cpirent.label.cpi_current">Current CPI</span>
                    <input type="number" step="0.1" min="0" name="cpi_current" value="295.4" required></label>
                <label><span data-i18n="view.cpirent.label.min">Floor on increase (%, 0 = none)</span>
                    <input type="number" step="0.1" min="0" name="min_increase_pct" value="2"></label>
                <label><span data-i18n="view.cpirent.label.max">Ceiling on increase (%, 0 = none)</span>
                    <input type="number" step="0.1" min="0" name="max_increase_pct" value="5"></label>
                <label><span data-i18n="view.cpirent.label.index">Index name</span>
                    <input type="text" name="index_label" value="CPI-U, U.S. city average"></label>
                <label><span data-i18n="view.cpirent.label.date">Effective date</span>
                    <input type="date" name="effective_date" value="2026-07-01" required></label>
                <label><span data-i18n="view.cpirent.label.statute">Statute citation (optional)</span>
                    <input type="text" name="statute_citation" value="" placeholder="${esc(t('view.cpirent.ph.statute'))}"></label>
            </form>
        </div>
        <div id="cpirent-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#cpirent-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            landlord_name: (fd.get('landlord_name') || '').trim(),
            tenant_name: (fd.get('tenant_name') || '').trim(),
            property_label: (fd.get('property_label') || '').trim(),
            base_rent_usd: Number(fd.get('base_rent_usd')) || 0,
            cpi_base: Number(fd.get('cpi_base')) || 0,
            cpi_current: Number(fd.get('cpi_current')) || 0,
            min_increase_pct: Number(fd.get('min_increase_pct')) || 0,
            max_increase_pct: Number(fd.get('max_increase_pct')) || 0,
            index_label: (fd.get('index_label') || '').trim(),
            effective_date: fd.get('effective_date'),
            state: (fd.get('state') || '').trim(),
            statute_citation: (fd.get('statute_citation') || '').trim(),
        };
        try {
            const doc = await api.calcCpiRentAdjustment(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.cpirent.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    form.addEventListener('input', () => { live(); });
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase()];
    if (doc.statutory_citation) lines.push(doc.statutory_citation);
    lines.push('');
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    let bound = '';
    if (doc.ceiling_applied) bound = ` <span class="muted small" data-i18n="view.cpirent.ceiling">(ceiling)</span>`;
    else if (doc.floor_applied) bound = ` <span class="muted small" data-i18n="view.cpirent.floor">(floor)</span>`;
    const el = mount.querySelector('#cpirent-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.cpirent.card.new">New annual rent</div>
                    <div class="value">${money(doc.new_rent_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.cpirent.card.increase">Increase</div>
                    <div class="value">${money(doc.dollar_increase_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.cpirent.card.applied">Applied increase</div>
                    <div class="value">${pct(doc.applied_increase_pct)}${bound}</div></div>
                <div class="card"><div class="label" data-i18n="view.cpirent.card.raw">Raw index change</div>
                    <div class="value">${pct(doc.raw_increase_pct)}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="cpirent-copy" type="button" data-i18n="view.cpirent.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="cpirent-download" type="button" data-i18n="view.cpirent.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#cpirent-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.cpirent.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.cpirent.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#cpirent-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'cpi-rent-adjustment.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/cpi_rent_adjustment.rs
//! CPI rent adjustment notice — the index-based escalation used in commercial
//! leases. Rent is reset by the ratio of a current consumer price index to a
//! base-period index, bounded by an optional collar (a floor and/or ceiling on
//! the percentage increase). This differs from the fixed-percentage compounding
//! in `rent_escalation`: the increase is driven by published CPI, not a constant
//! rate, and the collar clamps it. Drafting aid, not legal advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct CpiRentInput {
    pub landlord_name: String,
    pub tenant_name: String,
    #[serde(default)]
    pub property_label: String,
    /// Current annual rent before adjustment.
    pub base_rent_usd: f64,
    /// CPI level for the base period (lease commencement or last adjustment).
    pub cpi_base: f64,
    /// CPI level for the current period.
    pub cpi_current: f64,
    /// Floor on the percentage increase (0 = none). Rent rises by at least this.
    #[serde(default)]
    pub min_increase_pct: f64,
    /// Ceiling on the percentage increase (0 = uncapped).
    #[serde(default)]
    pub max_increase_pct: f64,
    /// Index name, for the notice (e.g. "CPI-U, U.S. city average").
    #[serde(default)]
    pub index_label: String,
    pub effective_date: String,
    pub state: String,
    #[serde(default)]
    pub statute_citation: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct CpiRentAdjustment {
    pub title: String,
    /// cpi_current ÷ cpi_base.
    pub index_ratio: f64,
    /// Raw percentage change implied by the index ratio.
    pub raw_increase_pct: f64,
    /// Percentage increase after applying the collar.
    pub applied_increase_pct: f64,
    /// True when the ceiling reduced the increase.
    pub ceiling_applied: bool,
    /// True when the floor raised the increase.
    pub floor_applied: bool,
    pub new_rent_usd: f64,
    pub dollar_increase_usd: f64,
    pub statutory_citation: String,
    pub clauses: Vec<DocClause>,
}

fn round4(x: f64) -> f64 {
    (x * 10_000.0).round() / 10_000.0
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

pub fn generate(i: &CpiRentInput) -> CpiRentAdjustment {
    let ratio = if i.cpi_base > 0.0 {
        i.cpi_current / i.cpi_base
    } else {
        1.0
    };
    let raw_pct = (ratio - 1.0) * 100.0;

    // Apply the ceiling first, then the floor, so a collar like 2%–5% pins the
    // increase into [min, max] regardless of which bound binds.
    let mut applied = raw_pct;
    let mut ceiling_applied = false;
    let mut floor_applied = false;
    if i.max_increase_pct > 0.0 && applied > i.max_increase_pct {
        applied = i.max_increase_pct;
        ceiling_applied = true;
    }
    if applied < i.min_increase_pct {
        applied = i.min_increase_pct;
        floor_applied = true;
    }

    let new_rent = cents(i.base_rent_usd * (1.0 + applied / 100.0));
    let increase = cents(new_rent - i.base_rent_usd);

    let property = if i.property_label.trim().is_empty() {
        "the Premises".to_string()
    } else {
        i.property_label.trim().to_string()
    };
    let index = if i.index_label.trim().is_empty() {
        "the Consumer Price Index".to_string()
    } else {
        i.index_label.trim().to_string()
    };

    let collar_desc = match (i.min_increase_pct > 0.0, i.max_increase_pct > 0.0) {
        (true, true) => format!(" The increase is collared between {:.2}% and {:.2}%.", i.min_increase_pct, i.max_increase_pct),
        (true, false) => format!(" The increase is floored at {:.2}%.", i.min_increase_pct),
        (false, true) => format!(" The increase is capped at {:.2}%.", i.max_increase_pct),
        (false, false) => String::new(),
    };

    let bound_note = if ceiling_applied {
        " (reduced to the ceiling)"
    } else if floor_applied {
        " (raised to the floor)"
    } else {
        ""
    };

    let calc_body = format!(
        "The index ratio is {} ÷ {} = {:.4}, an implied change of {:.4}%.{} The applied increase is {:.4}%{}. Base rent of {} becomes {}, an increase of {}.",
        i.cpi_current,
        i.cpi_base,
        round4(ratio),
        round4(raw_pct),
        collar_desc,
        round4(applied),
        bound_note,
        money(i.base_rent_usd),
        money(new_rent),
        money(increase)
    );

    let citation = i.statute_citation.trim();
    let governing = if citation.is_empty() {
        format!("This adjustment is made under the lease and the laws of the State of {}.", i.state)
    } else {
        format!("This adjustment is made under the lease and the laws of the State of {} ({}).", i.state, citation)
    };

    let clauses = vec![
        DocClause {
            heading: "Parties".into(),
            body: format!(
                "Landlord: {}\nTenant: {}\nPremises: {}\nEffective date: {}",
                i.landlord_name, i.tenant_name, property, i.effective_date
            ),
        },
        DocClause {
            heading: "1. Index Adjustment".into(),
            body: format!(
                "Rent is adjusted by reference to {}. The base-period index is {} and the current index is {}.{}",
                index, i.cpi_base, i.cpi_current, collar_desc
            ),
        },
        DocClause { heading: "2. Calculation".into(), body: calc_body },
        DocClause {
            heading: "3. New Rent".into(),
            body: format!(
                "Effective {}, the annual rent is {}, payable in equal monthly installments. The adjustment does not waive any other lease term.",
                i.effective_date, money(new_rent)
            ),
        },
        DocClause { heading: "4. Governing Law".into(), body: governing },
        DocClause {
            heading: "Signatures".into(),
            body: format!(
                "Landlord: ____________________  Date: __________\n{}",
                i.landlord_name
            ),
        },
    ];

    CpiRentAdjustment {
        title: "CPI Rent Adjustment Notice".into(),
        index_ratio: round4(ratio),
        raw_increase_pct: round4(raw_pct),
        applied_increase_pct: round4(applied),
        ceiling_applied,
        floor_applied,
        new_rent_usd: new_rent,
        dollar_increase_usd: increase,
        statutory_citation: citation.to_string(),
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> CpiRentInput {
        CpiRentInput {
            landlord_name: "Tower Owners LP".into(),
            tenant_name: "Office Tenant LLC".into(),
            property_label: "Floor 7".into(),
            base_rent_usd: 50_000.0,
            cpi_base: 280.0,
            cpi_current: 295.4,
            min_increase_pct: 2.0,
            max_increase_pct: 5.0,
            index_label: "CPI-U".into(),
            effective_date: "2026-07-01".into(),
            state: "Delaware".into(),
            statute_citation: String::new(),
        }
    }

    #[test]
    fn ceiling_caps_large_increase() {
        let d = generate(&base());
        assert!(close(d.index_ratio, 1.055));
        assert!(close(d.raw_increase_pct, 5.5));
        assert!(close(d.applied_increase_pct, 5.0));
        assert!(d.ceiling_applied);
        assert!(!d.floor_applied);
        assert!(close(d.new_rent_usd, 52_500.0));
        assert!(close(d.dollar_increase_usd, 2_500.0));
    }

    #[test]
    fn no_collar_uses_raw_change() {
        let d = generate(&CpiRentInput {
            cpi_current: 287.0,
            min_increase_pct: 0.0,
            max_increase_pct: 0.0,
            ..base()
        });
        assert!(close(d.raw_increase_pct, 2.5));
        assert!(close(d.applied_increase_pct, 2.5));
        assert!(close(d.new_rent_usd, 51_250.0));
    }

    #[test]
    fn floor_lifts_when_cpi_falls() {
        let d = generate(&CpiRentInput { cpi_current: 275.0, ..base() });
        // Raw is negative; floor of 2% applies.
        assert!(d.raw_increase_pct < 0.0);
        assert!(close(d.applied_increase_pct, 2.0));
        assert!(d.floor_applied);
        assert!(!d.ceiling_applied);
        assert!(close(d.new_rent_usd, 51_000.0));
    }

    #[test]
    fn within_collar_unbounded() {
        let d = generate(&CpiRentInput { cpi_current: 290.0, ..base() });
        // Ratio ~1.0357 → ~3.57%, inside 2–5%.
        assert!(!d.ceiling_applied);
        assert!(!d.floor_applied);
        assert!(close(d.applied_increase_pct, d.raw_increase_pct));
    }

    #[test]
    fn zero_base_index_no_change() {
        let d = generate(&CpiRentInput { cpi_base: 0.0, min_increase_pct: 0.0, max_increase_pct: 0.0, ..base() });
        assert!(close(d.index_ratio, 1.0));
        assert!(close(d.new_rent_usd, 50_000.0));
    }

    #[test]
    fn statute_citation_echoed() {
        let d = generate(&CpiRentInput { statute_citation: "lease § 5.1".into(), ..base() });
        assert_eq!(d.statutory_citation, "lease § 5.1");
        assert!(d.clauses.iter().any(|c| c.body.contains("lease § 5.1")));
    }
}
